src/libopensc/reader-cryptotokenkit.m

```diff
@@ -0,0 +1,426 @@
+			reply:^(NSData *response, NSError *error) {
+		if (response) {
+			rsize = [response length];
+			rbuf = malloc(rsize);
+			if (!rbuf)
+				r = SC_ERROR_OUT_OF_MEMORY;
+			else
+				memcpy(rbuf, (unsigned char*) [response bytes], rsize);
+		}
+		if (r == SC_SUCCESS) {
+			if (error) {
+				NSLog(@"Error transmitting to card <%@>", error);
+				r = SC_ERROR_TRANSMIT_FAILED;
+			} else {
+				r = SC_SUCCESS;
+			}
+		}
+		dispatch_semaphore_signal(sema);
+	}];
+	dispatch_semaphore_wait(sema, DISPATCH_TIME_FOREVER);
+	if (r != SC_SUCCESS)
+		goto err;
+
+	sc_apdu_log(reader->ctx, SC_LOG_DEBUG_NORMAL, rbuf, rsize, 0);
+	r = sc_apdu_set_resp(reader->ctx, apdu, rbuf, rsize);
+
+err:
+	if (sbuf != NULL) {
+		sc_mem_clear(sbuf, ssize);
+		free(sbuf);
+	}
+	if (rbuf != NULL) {
+		sc_mem_clear(rbuf, rsize);
+		free(rbuf);
+	}
+
+	LOG_FUNC_RETURN(reader->ctx, r);
+}
+
+int cryptotokenkit_use_reader(sc_context_t *ctx, void *pcsc_context_handle, void *pcsc_card_handle)
+{
+	int r;
+	struct cryptotokenkit_private_data *priv;
+	sc_reader_t *reader = NULL;
+	scconf_block *conf_block = NULL;
+	TKSmartCardSlot* tksmartcardslot = (__bridge TKSmartCardSlot *)(pcsc_context_handle);
+	TKSmartCard* tksmartcard = (__bridge TKSmartCard *)(pcsc_card_handle);
+	const char* utf8String;
+
+	if (!pcsc_context_handle) {
+		if (!pcsc_card_handle)
+			return SC_ERROR_INVALID_ARGUMENTS;
+		tksmartcardslot = tksmartcard.slot;
+	}
+
+	if ((reader = calloc(1, sizeof(sc_reader_t))) == NULL) {
+		r = SC_ERROR_OUT_OF_MEMORY;
+		goto err;
+	}
+	if ((priv = calloc(1, sizeof(struct cryptotokenkit_private_data))) == NULL) {
+		r = SC_ERROR_OUT_OF_MEMORY;
+		goto err;
+	}
+	[priv->tksmartcard autorelease];
+	priv->tksmartcard = [tksmartcard retain];
+	[priv->tksmartcardslot autorelease];
+	priv->tksmartcardslot = [tksmartcardslot retain];
+
+	reader->drv_data = priv;
+	reader->ops = &cryptotokenkit_ops;
+	reader->driver = &cryptotokenkit_reader_driver;
+	utf8String = [tksmartcardslot.name UTF8String];
+	if ((reader->name = strdup(utf8String)) == NULL) {
+		r = SC_ERROR_OUT_OF_MEMORY;
+		goto err;
+	}
+
+	/* By testing we found that maxInputLength/maxOutputLength are
+	 * most likely initialized badly. We still take this value as is
+	 * and leave it up to the user to overwrite the reader's
+	 * capabilities */
+	reader->max_send_size = tksmartcardslot.maxInputLength;
+	reader->max_recv_size = tksmartcardslot.maxOutputLength;
+
+	conf_block = sc_get_conf_block(ctx, "reader_driver", "cryptotokenkit", 1);
+	if (conf_block) {
+		reader->max_send_size = scconf_get_int(conf_block, "max_send_size", reader->max_send_size);
+		reader->max_recv_size = scconf_get_int(conf_block, "max_recv_size", reader->max_recv_size);
+	}
+
+	/* attempt to detect protocol in use T0/T1/RAW */
+	ctk_set_proto(reader);
+
+	r = _sc_add_reader(ctx, reader);
+
+err:
+	if (r != SC_SUCCESS) {
+		free(priv);
+		if (reader != NULL) {
+			if (reader->name)
+				free(reader->name);
```

Comment:
Its perfectly safe to call free on NULL pointer
C standard, 7.20.3.2/2 from ISO-IEC 9899
If ptr is a null pointer, no action occurs.